src/uitext.ml

```diff
@@ -86,6 +86,63 @@ let setupTerminal() =
     with Unix.Unix_error _ ->
       restoreTerminal ()
 
+let colorMode =
+  Prefs.createBoolWithDefault "color"
+    "!use color output for text UI (true/false/default)"
+    ("When set to \\verb|true|, this flag enables color output in "
+     ^ "text mode user interface. When set to \\verb|false|, all "
+     ^ "color output is disabled. Default is to enable color if "
+     ^ "the NO_COLOR environment variable is not set.")
+
+let colorEnabled = ref false
+
+let setColorPreference () =
+  let envOk = ((Sys.getenv_opt "NO_COLOR") = None) in
+  colorEnabled := ((Unix.isatty Unix.stdin) && (Unix.isatty Unix.stderr) &&
+    (match Prefs.read colorMode with
+      `True    -> true
+    | `False   -> false
+    | `Default -> envOk))
+
+let color t =
+  if not !colorEnabled then "" else
+  match t with
+    `Reset       -> "\o033[0m"
+  | `Success     -> "\o033[1;32m"
+  | `Information -> "\o033[1;34m"
+  | `Warning     -> "\o033[1;33m"
+  | `Failure     -> "\o033[1;31m"
+  | `AError      -> "\o033[31m"
+  | `ASkip       -> "\o033[1;35m"
+  | `ALtoRf      -> "\o033[1;32m"
+  | `ALtoRt      -> "\o033[1;33m"
+  | `ARtoLf      -> "\o033[1;34m"
+  | `ARtoLt      -> "\o033[1;33m"
+  | `AMerge      -> "\o033[1;36m"
+  | `DiffHead    -> "\o033[1m"
+  | `DiffAdd     -> "\o033[32m"
+  | `DiffDel     -> "\o033[31m"
+  | `DiffLoc     -> "\o033[36m"
+  | _            -> ""
+
+let lineRegexp = Str.regexp "^"
+
+let colorDiff text =
+  let result = ref "" in
+  let p = Str.full_split lineRegexp text in
+  Safelist.iter (fun t ->
+              result := !result ^
```

Comment:
I know this is not the OCaml way. Initially, I was concerned about the recursion because the diffs could be very long. But I hadn't realized then that Safelist implements tail-recursive functions.
What is the preferred way? Is `String.concat "" (Safelist.map (...) p)` reasonable?